### PR TITLE
SF-2213 Disable chapter audio upload when offline

### PR DESCRIPTION
**Pull request by marksvc**
Video showing behaviour:
[offline-audio-upload.webm](https://github.com/sillsdev/web-xforge/assets/7265309/1309cccc-34d6-441a-a323-c1759c2e5bc9)

This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)

### CODE REVIEW

**Review by marksvc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @github-code-scanning[bot])

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts` line 5 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2107#gh-1346427299:-NfwP5uQ-wCQsNg1puYM:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/b2ed1f372963d179e3b9de04cc46c53ba6a67eb4/src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts#L5)):*
<details><summary><i>Previously, github-code-scanning[bot] wrote…</i></summary><blockquote>

## Unused variable, import, function or class

Unused import By.

[Show more details](https://github.com/sillsdev/web-xforge/security/code-scanning/646)
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts` line 18 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2107#gh-1346427305:-NfwP69mE98WyG-1x6MT:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/b2ed1f372963d179e3b9de04cc46c53ba6a67eb4/src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts#L18)):*
<details><summary><i>Previously, github-code-scanning[bot] wrote…</i></summary><blockquote>

## Unused variable, import, function or class

Unused import PwaService.

[Show more details](https://github.com/sillsdev/web-xforge/security/code-scanning/647)
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts` line 27 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2107#gh-1346427308:-NfwP6a72aonnuSAAWAC:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/b2ed1f372963d179e3b9de04cc46c53ba6a67eb4/src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts#L27)):*
<details><summary><i>Previously, github-code-scanning[bot] wrote…</i></summary><blockquote>

## Unused variable, import, function or class

Unused import of.

[Show more details](https://github.com/sillsdev/web-xforge/security/code-scanning/648)
</blockquote></details>

Done.

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: all files reviewed, 5 unresolved discussions (waiting on @github-code-scanning[bot] and @marksvc)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.scss` line 143 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-NfwjRPRD06SbPf773Ij:-NfwjRPRD06SbPf773Ik:b-46mta) ([raw file](https://github.com/sillsdev/web-xforge/blob/34be6845180d03aa0a59b8ee0ca3e37aa5bf6be0/src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.scss#L143)):*
> ```SCSS
> .offline-text {
>   padding-left: 16px;
>   padding-top: 16px;
> ```

The text doesn't quite look like it is aligned to the left of the dialog because of this padding. Do we need it? In RTL mode it is right aligned. We should be consistent whichever direction we choose.



_Code quote:_
```SCSS
.offline-text {
  padding-left: 16px;
  padding-top: 16px;
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/checking_en.json` line 216 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-Nfwiu5eBsIUzlzMVgqy:-Nfwiu5eBsIUzlzMVgqz:bhau330) ([raw file](https://github.com/sillsdev/web-xforge/blob/34be6845180d03aa0a59b8ee0ca3e37aa5bf6be0/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/checking_en.json#L216)):*
> ```JSON
>     "no_audio_file_uploaded": "No audio file uploaded",
>     "no_timing_data_uploaded": "No timing data uploaded",
>     "not_while_offline_connect_to_continue": "This can not be done while offline. Connect to the internet to continue.",
> ```

I like being more specific when possible. Something like "Audio files cannot be saved while offline. Connect to the internet to continue."



_Code quote:_
```JSON
   "not_while_offline_connect_to_continue": "This can not be done while offline. Connect to the internet to continue.",
```

**Review by marksvc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: all files reviewed, 5 unresolved discussions (waiting on @github-advanced-security[bot] and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.scss` line 143 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-NfwjRPRD06SbPf773Ij:-NhTaD3EE1B67MamKWxu:b-f7o1da) ([raw file](https://github.com/sillsdev/web-xforge/blob/34be6845180d03aa0a59b8ee0ca3e37aa5bf6be0/src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.scss#L143)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

The text doesn't quite look like it is aligned to the left of the dialog because of this padding. Do we need it? In RTL mode it is right aligned. We should be consistent whichever direction we choose.
</blockquote></details>

I was trying to indent it a bit, but I don't remember why :). Removed.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/checking_en.json` line 216 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-Nfwiu5eBsIUzlzMVgqy:-NhTatb22j9wmTQ2pn2u:b-leqhir) ([raw file](https://github.com/sillsdev/web-xforge/blob/34be6845180d03aa0a59b8ee0ca3e37aa5bf6be0/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/checking_en.json#L216)):*
Thank you. How about, 
> Audio files cannot be uploaded while offline. Connect to the internet to continue.

**Review by RaymondLuong3 — APPROVED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: all files reviewed, 3 unresolved discussions (waiting on @github-advanced-security[bot])

**Comment by marksvc** [comment]
I further changed it when I saw the nice `<app-notice>` component. To add margin atop the app-notice element, I changed notice.component to be `display: block` instead of inline. notice.component is used in a few places in SF. It looks like they should all be fine with display set to block.

Screenshots showing the new look and message:

Online:
image

Offline:
image

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@
   getAudioBlob
 } from 'xforge-common/test-utils';
 import { UICommonModule } from 'xforge-common/ui-common.module';
+import { BehaviorSubject } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/chapter-audio-dialog/chapter-audio-dialog.component.spec.ts:26]: Done.

**Review by Nateowami — COMMENTED** [review]
Overall looks good. A couple thoughts.

Personally I think the red background (as opposed to outline with red text) is over the top. Elsewhere the offline messages are much more subdued:



Then again, in this dialog the *only thing* you can do is unavailable, so maybe it's warranted? 

> Reviewed 1 of 6 files at r3, 2 of 4 files at r4, 3 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-:-NhYu0_HB4KwJRxplzPf:b-m2hevj)* status: 6 of 7 files reviewed, 5 unresolved discussions (waiting on @github-advanced-security[bot], @marksvc, and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss` line 4 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-NhYrW3x6ptVUgG_e26e:-NhYrW3x6ptVUgG_e26f:bd4rxov) ([raw file](https://github.com/sillsdev/web-xforge/blob/05c68aeb5230dba10fca56593a8ed9bf0072190b/src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss#L4)):*
> ```SCSS
> 
> :host {
>   display: block;
> ```

Did you check everywhere this component is used to make sure this isn't causing problems elsewhere? This is a good change, but makes the most sense to make at the moment the component is created. I doubt it will cause any problems elsewhere, but we can't be sure without checking.

Also, maybe you could update `src/SIL.XForge.Scripture/ClientApp/src/xforge-common/feature-flags/feature-flags-dialog.component.scss` to remove `display: block;` from the styles there, since it's now redundant?

**Review by Nateowami — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: 6 of 7 files reviewed, 5 unresolved discussions (waiting on @github-advanced-security[bot], @marksvc, and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss` line 4 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-NhYrW3x6ptVUgG_e26e:-NhYvtk4DSwYPjwbXloT:b-8cyhek) ([raw file](https://github.com/sillsdev/web-xforge/blob/05c68aeb5230dba10fca56593a8ed9bf0072190b/src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss#L4)):*
Actually you addressed this very question:
> notice.component is used in a few places in SF. It looks like they should all be fine with display set to block.

Thanks for remembering to look at that. I saw that the PR had been updated very recently and mistakenly thought it was new and didn't read through the comments.

**Review by marksvc — COMMENTED** [review]
It is pretty loud. I have set the message to use outline instead. How about this? 



> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-:-NhcX6RwDxeYGH8_IF-M:b-v8fj0u)* status: 6 of 7 files reviewed, 5 unresolved discussions (waiting on @github-advanced-security[bot], @Nateowami, and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss` line 4 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-NhYrW3x6ptVUgG_e26e:-NhcU5uB-DZ2Moquifdh:blz4asf) ([raw file](https://github.com/sillsdev/web-xforge/blob/05c68aeb5230dba10fca56593a8ed9bf0072190b/src/SIL.XForge.Scripture/ClientApp/src/app/shared/notice/notice.component.scss#L4)):*
I looked at most of the places in the browser that I knew how to easily get to. More comprehensively, all seemed like it would be fine from looking at their containing html/templates. 

> makes the most sense to make at the moment the component is created.

Yes, this could have unfortunate consequences depending on how the component is being used. But it looks like we are using it on its own line each time.

> feature-flag-dialog

Done!

**Comment by marksvc** [comment]
@RaymondLuong3 I've made some changes since your approval, if you can have a look at what's new. Thanks!

**Review by Nateowami — COMMENTED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107)* status: 7 of 8 files reviewed, 4 unresolved discussions (waiting on @github-advanced-security[bot] and @RaymondLuong3)

**Review by RaymondLuong3 — APPROVED** [review]
Thanks for the changes. I think using the app-notice may be overkill, but I don't see any issues with it. I approve.

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2107#-:-Nhh87xo9eOHuk5cHL61:b-d6tivr)* status: all files reviewed, 4 unresolved discussions (waiting on @github-advanced-security[bot])

**Comment by marksvc** [comment]
@RaymondLuong3 , @Nateowami: Hmm. I'm okay with using a different look for the message. It may be good to choose a unified style for our various "Note: You can't do some or all of this while offline" messages that are here and there in our application. Some are using `<app-notice`, and some are using `<span class="offline-text"`.

**Pull request merged**